Userland/Utilities/tsort.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022, Eli Youngs <eli.m.youngs@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/HashMap.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/Stream.h>
+#include <LibCore/System.h>
+#include <LibMain/Main.h>
+
+enum NodeStatus {
+    NotSeen,
+    Seen,
+    Prioritized,
+};
+
+struct Node {
+    StringView name;
+    HashTable<Node*> ancestors;
+    NodeStatus status;
+
+    unsigned hash() const { return name.hash(); }
+};
+
+using NodeMap = HashMap<StringView, Node*>;
+using NodeStack = Vector<Node*>;
+
+static Node* get_or_create_node(NodeMap& map, StringView const key)
+{
+    Optional<Node*> maybe_child_node = map.get(key);
+    if (maybe_child_node.has_value())
+        return maybe_child_node.release_value();
+
+    Node* new_node = new Node { key, HashTable<Node*> {}, NodeStatus::NotSeen };
+    map.set(key, new_node);
+    return map.get(key).release_value();
+}
+
+static void handle_cycle(NodeStack* stack, Node* duplicated_node)
+{
+    // Report on a cycle by moving down the stack of dependencies, logging every node
+    // between the implicit top of the stack (represented by duplicate_node) and that
+    // node's first appearance.
+
+    warnln("tsort: The following nodes form a cycle");
+    for (auto it = stack->rbegin(); it != stack->rend(); ++it) {
+        auto* node = *it;
+        warnln("tsort: {}", node->name);
+        node->status = NodeStatus::NotSeen;
+        if (node == duplicated_node)
+            return;
+    }
+}
+
+static void prioritize_nodes(Node* start)
+{
+    // Prioritize (topologically sort) a subset of a directed graph using a depth first
+    // search. The "deepest" nodes are the earliest ancestors of all other nodes and
+    // have no dependencies. To avoid a stack overflow when processing deep dependency
+    // chains, this function does not call itself recursively. Instead, the recursive
+    // algorithm is implemented on a heap-allocated stack.
+
+    static auto* stack = new NodeStack;
```

Comment:
NodeStack can be stored on stack, as a normal variable. `Vector` itself allocates data on the heap, avoiding stack overflow.
```suggestion
    static NodeStack stack;
```